### PR TITLE
update matsim

### PR DESCRIPTION
**Pull request by nkuehnel**
- **change java version to 21**
- **explicitly configure java version in maven compiler plugin**
- **build(deps): bump com.google.errorprone:error_prone_annotations**
- **build(deps): bump org.checkerframework:checker-qual**
- **build(deps): bump com.google.guava:guava from 33.1.0-jre to 33.2.0-jre**
- **replace population comparison in tests and deprecate by byte comparison of plan files (#3247); add attributes comparison**

### CODE REVIEW

File: matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java [hunk 1]
```diff
@@ -3,127 +3,185 @@
...
-				return false;
-			}
-		} else {
-			throw new RuntimeException ("Unexpected PlanElement");
-		}
-		return true;
-	}
+public class PopulationComparison {
+
+    private final static double DEFAULT_DELTA = 1e-10;
```
Comment (mfrawley-moia) [inline matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java:15]: @nkuehnel didn't you say that there is a default epsilon in Matsim which has this same value? Should we re-use that one or use this?

File: matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java [hunk 2]
```diff
@@ -3,127 +3,185 @@
...
+
+            if (!person1.getId().equals(person2.getId())) {
+                log.warn("");
+                log.warn("persons out of sequence p1: " + person1.getId() + " | p2: " + person2.getId());
+                result = Result.notEqual;
+                continue;
+            }
+
+            if (!AttributesComparison.equals(person1.getAttributes(), person2.getAttributes())) {
+                log.warn("");
```
Comment (mfrawley-moia) [inline matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java:60]: empty output, is it a way of injecting newlines into logs for readability?

File: matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java [hunk 3]
```diff
@@ -3,127 +3,185 @@
...
+            for (int i = 0; i < nElements; i++) {
+                if (!equals(planElements.get(i), planElements2.get(i), delta)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    /* Warning: This is NOT claimed to be correct. (It isn't.)
```
Comment (mfrawley-moia) [inline matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java:150]: What are the known issues? Should we have a test to show it works for the existing use-case?

File: matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java (hunk 1 above)
Comment (nkuehnel) [reply matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java:15]: @mfrawley-moia there is, but is in only visible in the test scope of maven not in compile. The MatsimTestUtils cant be accessed here

File: matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java (hunk 2 above)
Comment (nkuehnel) [reply matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java:60]: yes I copied that from the previous behavior of this class

File: matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java (hunk 3 above)
Comment (nkuehnel) [reply matsim/src/main/java/org/matsim/core/population/routes/PopulationComparison.java:150]: this is a comment i copied over from a matsim test where this method used to live (the previous population comparison didnt test the plan elements)

**Review by mfrawley-moia — APPROVED** [review]

**Pull request merged**